### PR TITLE
34: replace url crate with peg parser based on RFC 3986

### LINKED ISSUE #34: rewrite ethr parser
### Is your feature request related to a problem?

The `DidUrl` parser uses [URL crate](https://crates.io/crates/url) to assist parsing but it's problematic as [specified here](https://github.com/xmtp/didethresolver/blob/main/src/types/did_url.rs#L148-L151). [uri crate](https://crates.io/crates/uri) doesn't work for our case as well.

### Describe the solution to the problem

So we will need to write our own parser grammar in order to properly parse a did url with optional path, query and fragment like this:
"did:ethr:0xb9c5714089478a327f09197987f16f9e5d936e8a/path?query=q#fragment"

### Describe the uses cases for the feature

We may need to accommodate other did format, but for now the feature would be more than enough.

### Additional details

[lyn](https://docs.rs/lyn/latest/lyn/struct.Scanner.html) might be a good reference.

### PR DESCRIPTION
**Pull request by jac18281828**
closes #34 

- adding path + query + fragment
- test coveragee
- parses query as a vec of string tuple

general cleanup, make DidUrl immutable by removing setter

fix rustdoc

### CODE REVIEW

**Comment by insipx** [comment]
🧹 `url` 🧹

nice! I like the contextual errors for parsing, I hope it will make it easier on users if they get the DID wrong

**Review by insipx — APPROVED** [review]

**Review by 37ng — COMMENTED** [review]
Would it make sense to use map instead of an array for queries? It's more efficient to search keys.
Also, would it make sense to test immutability? i.e. checking that original object is not modified after calling `with_[query/path/attribute/...]`?

File: lib/src/types/did_parser.rs [hunk 1]
```diff
@@ -28,30 +53,94 @@ peg::parser! {
         ///   });
         /// ```
         pub rule ethr_did() -> Did
-            = method:method() ":" network:network()? account:account() { Did { method, network: network.unwrap_or_default(), account } }
+            = did() ":" method:method() ":" network:(chain_id() / network())? account:account() { Did { method, network: network.unwrap_or_default(), account } }
```
Comment (37ng) [inline lib/src/types/did_parser.rs:56]: In the old version, the network will be defaulted to `mainnet` when eluded. I assume we don't want to keep that logic now? I'm fine with that tho

**Comment by jac18281828** [comment]
> Would it make sense to use map instead of an array for queries? It's more efficient to search keys. Also, would it make sense to test immutability? i.e. checking that original object is not modified after calling `with_[query/path/attribute/...]`?

It's a good idea to use the map but it will not work because the names and values are not distinct in the specification. A query could have a=a, a=a, a=a and be valid syntax its application dependent what the meaning of that is.   Imagine the protocol is 'execute this instruction'.  

The performance of an array vs a map for 2-3 elements is probably actually better for lookup in a typical application where every parameter will have meaning.      Since we are using 1 at the moment we can probably ignore this issue.

The test of immutability should be guaranteed by rust ownership rules, but I will add a check for sanity.

**Comment by 37ng** [comment]
> The test of immutability should be guaranteed by rust ownership rules, but I will add a check for sanity.

In unit tests, imo it's crucial not to depend on implementation specifics. For functions with_*** functions, they should ideally return a new object rather than altering the original one. Our tests must verify this behavior directly, avoiding assumptions about Rust's clone() method or omitting such tests.

Agreed with the query data structure, thanks for explaining.

**Comment by jac18281828** [comment]
Please take a last look.   I believe it's ready to go.  I updated with some cleanup:

* removed tiny-keccak
* added parsing support with a test for public key did....I suspect pks are still not working and we need to add a ticket to test that and verify it is all good.
* cleaned up account creation
* added coverage for codecov complaints

File: lib/src/types/did_url.rs [hunk 2]
```diff
@@ -1,20 +1,30 @@
...
+use ethers::{types::Address, utils::keccak256};
+
+use super::parse_ethr_did_url;
 use crate::error::DidError;
 
+/**
+ * Converts a public key to an ethereum address.  The last 20 bytes
+ * of the keccak256 hash of the public key is returned
+ */
+fn public_key_to_address(key: &[u8]) -> String {
```
Comment (37ng) [inline lib/src/types/did_url.rs:15]: Would this [function](https://docs.ethers.org/v5/api/utils/address/#utils-computeAddress) from ethers be helpful?

File: lib/src/types/did_url.rs [hunk 3]
```diff
@@ -54,7 +64,37 @@ impl ToString for Account {
...
         match self {
             Account::Address(addr) => format!("0x{}", hex::encode(addr.as_bytes())),
-            Account::HexKey(key) => format!("0x{}", hex::encode(key)),
+            Account::HexKey(key) => hex::encode(key).to_string(),
+        }
+    }
+}
+
+impl Account {
+    pub fn from_hex_key(key: Vec<u8>) -> Self {
```
Comment (37ng) [inline lib/src/types/did_url.rs:73]: maybe use From would be more idiomatic?
```rust
impl From<HexKey/Address/&str> for Account {
    // implementation
}
```

File: lib/src/types/did_url.rs (hunk 2 above)
Comment (jac18281828) [reply lib/src/types/did_url.rs:15]: The reason I didn't use that is it requires a wallet.    It felt ceremonial to instantiate all that when we don't require the wallet ourselves.

File: lib/src/types/did_url.rs (hunk 3 above)
Comment (jac18281828) [reply lib/src/types/did_url.rs:73]: 💯💯 thank you this feedback is so helpful, I did this method kind of fast and forgot to even think about From 🙈

**Pull request merged**